ArduCopter/Copter.cpp

```diff
@@ -688,12 +688,73 @@ void Copter::three_hz_loop()
     low_alt_avoidance();
 }
 
+// ap_value calculates a 32-bit bitmask representing various pieces of
+// state about the Copter.  It replaces a global variable which was
+// used to track this state.
+uint32_t Copter::ap_value() const
+{
+    enum class AP_Bit {
+        // unused1 = (1U << 0),
+        // unused_was_simple_mode = (1U<<1): 2, // 1,2
+        PRE_ARM_RC_CHECK = (1U << 3),           // true if rc input pre-arm checks have been completed successfully
+        PRE_ARM_CHECK = (1U << 4),              // true if all pre-arm checks (rc, accel calibration, gps lock) have been performed
+        AUTO_ARMED = (1U << 5),                 // stops auto missions from beginning until throttle is raised
+        LOGGING_STARTED = (1U << 6),            // true if logging has started
+        LAND_COMPLETE = (1U << 7),              // true if we have detected a landing
+        NEW_RADIO_FRAME = (1U << 8),            // Set true if we have new PWM data to act on from the Radio
+        // USB_CONNECTED_UNUSED = (1U << 9),    // UNUSED
+        // RC_RECEIVER_PRESENT_UNUSED = (1U << 10), // UNUSED
+        COMPASS_MOT = (1U << 11),               // true if we are currently performing compassmot calibration
+        MOTOR_TEST = (1U << 12),                // true if we are currently performing the motors test
+        INITIALISED = (1U << 13),               // true once the init_ardupilot function has completed.  Extended status to GCS is not sent until this completes
+        LAND_COMPLETE_MAYBE = (1U << 14),       // true if we may have landed (less strict version of land_complete)
+        THROTTLE_ZERO = (1U << 15),             // true if the throttle stick is at zero, debounced, determines if pilot intends shut-down when not using motor interlock
+        // SYSTEM_TIME_SET_UNUSED = (1U << 16), // true if the system time has been set from the GPS
+        GPS_GLITCHING = (1U << 17),             // true if GPS glitching is affecting navigation accuracy
+        USING_INTERLOCK = (1U << 20),           // aux switch motor interlock function is in use
+        LAND_REPO_ACTIVE = (1U << 21),          // true if the pilot is overriding the landing position
+        MOTOR_INTERLOCK_SWITCH = (1U << 22),    // true if pilot is requesting motor interlock enable
+        IN_ARMING_DELAY = (1U << 23),           // true while we are armed but waiting to spin motors
+        INITIALISED_PARAMS = (1U << 24),        // true when the all parameters have been initialised. we cannot send parameters to the GCS until this is done
+        // unused3 = (1U << 25),                // was compass_init_location; true when the compass's initial location has been set
+        // unused2 = (1U << 26),                // aux switch rc_override is allowed
+        ARMED_WITH_AIRMODE_SWITCH = (1U << 27), // we armed using a arming switch
+        PREC_LAND_ACTIVE = (1U << 28),          // true if precland is active
+    };
+
+    uint32_t ret = 0;
+
```

Comment:
```
uint32_t bmask = 0;
const bool *b = (const uint8_t *)&ap;
for (uint16_t i=0; i<sizeof(ap)) {
 if (b[i]) {
  bmask |= 1U<<i;
 }
}
```

Reply:
This doesn't give the same result: there are gaps in the bitmask (e.g. the first 3 bits are unused).

Reply:
> This doesn't give the same result: there are gaps in the bitmask (e.g. the first 3 bits are unused).

Yep, sorry, we didn't capture all of the fun on the call last night - tridge made it abundantly clear he was only sketching :-)

I did actually implement it while on the call last night, but discovered the more serious issue that our *existing* implementation is flawed - if you have a look at the current bitmask commnets you will note there's a gap between 17 and 20.  I need to do some archeology to work out what the correct fix there is.

I've force-pushed this branch (complete with the debug which showed this problem - I was expecting iteration up to 28 but only got to 26)

Reply:
This is cute but is totally undefined behavior.

Reply:
> This is cute but is totally undefined behavior.

Oh?  How so?  Pretty sure we need to `PACK` this structure, but past that - do you see other undefined behaviour?

Reply:
Explained e.g. [here](https://wiki.sei.cmu.edu/confluence/display/c/ARR37-C.+Do+not+add+or+subtract+an+integer+to+a+pointer+to+a+non-array+object) . 

Each member is a separate object, which is an array of length 1, and simply generating a pointer more than 1 past the end of an array is undefined. `PACK` would make it clearer, and it's probably okay in practice, and other parts of the code use a similar pattern, but it's still undefined.

Reply:
A more C++ solution would be to have methods that return references so you could do e.g. `ap.motor_interlock_switch() = true`.

Reply:
> Explained e.g. [here](https://wiki.sei.cmu.edu/confluence/display/c/ARR37-C.+Do+not+add+or+subtract+an+integer+to+a+pointer+to+a+non-array+object) .
> 
> Each member is a separate object, which is an array of length 1, and simply generating a pointer more than 1 past the end of an array is undefined. `PACK` would make it clearer, and it's probably okay in practice, and other parts of the code use a similar pattern, but it's still undefined.

I'm pretty sure the packed attribute makes it just fine and defined.  As you say, if it doesn't then the hundreds of places we checksum by iterating across uint8_t-*-cast structures is going to be a problem :-)

Reply:
It's about the pointer arithmetic, not the packing. And I don't like those places either ;)

A union would be another idea I think.

Reply:
New structure is in place.